File: mybatis-generator-core/src/main/java/com/wvkity/mybatis/code/make/jdbc/JdbcTemplate.java
package com.wvkity.mybatis.code.make.jdbc;

import com.alibaba.fastjson.JSON;
import com.alibaba.fastjson.JSONArray;
import lombok.extern.log4j.Log4j2;

import java.sql.Connection;
import java.sql.DatabaseMetaData;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.ResultSetMetaData;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * JDBC工具
 * @author wvkity
 */
@Log4j2
public final class JdbcTemplate {
    private JdbcTemplate() {
    }

    /**
     * 执行更新操作
     * @param connection {@link Connection}
     * @param sql        SQL语句
     * @param args       参数
     * @return 受影响行数
     */
    public static int update(final Connection connection, final String sql, final Object... args) throws SQLException {
        try (final PreparedStatement statement = connection.prepareStatement(sql)) {
            if (args != null && args.length > 0) {
                for (int i = 0, size = args.length; i < size; i++) {
                    statement.setObject((i + 1), args[i]);
                }
            }
            return statement.executeUpdate();
        } catch (SQLException e) {
            log.error("`{}`执行失败: {}", sql, e.getMessage(), e);
            throw e;
        }
    }

    /**
     * 查询列表
     * @param clazz      指定返回类型
     * @param connection {@link Connection}
     * @param sql        SQL
     * @param args       参数
     * @param <T>        泛型类型
     * @return 列表
     * @throws SQLException 异常信息
     */
    public static <T> List<T> list(final Class<T> clazz, final Connection connection,
                                   final String sql, final Object... args) throws SQLException {
        final List<Map<String, Object>> result = list(connection, sql, args);
        if (!result.isEmpty()) {
            return JSONArray.parseArray(JSON.toJSONString(result), clazz);
        }
        return new ArrayList<>(0);
    }

    /**
     * 查询列表
     * @param connection {@link Connection}
     * @param sql        SQL
     * @param args       参数
     * @return 列表
     * @throws SQLException 异常信息
     */
    public static List<Map<String, Object>> list(final Connection connection, final String sql,
                                                 final Object... args) throws SQLException {
        return select(connection, sql, args);
    }

    /**
     * 查询
     * @param connection {@link Connection}
     * @param sql        SQL
     * @param args       参数
     * @return 多个值
     * @throws SQLException 异常信息
     */
    private static List<Map<String, Object>> select(final Connection connection, final String sql,
                                                    final Object... args) throws SQLException {
        try (final PreparedStatement statement = connection.prepareStatement(sql)) {
            if (args != null && args.length > 0) {
                for (int i = 0, size = args.length; i < size; i++) {
                    statement.setObject((i + 1), args[i]);
                }
            }
            final List<Map<String, Object>> result = new ArrayList<>();
            try (final ResultSet rs = statement.executeQuery()) {
                while (rs.next()) {
                    final Map<String, Object> rest = new HashMap<>();
                    final ResultSetMetaData metaData = rs.getMetaData();
                    for (int i = 1, size = metaData.getColumnCount(); i <= size; i++) {
                        final String key = metaData.getColumnName(i);
                        rest.put(key, rs.getObject(key));
                    }
                    result.add(rest);
                }
            }
            return result;
        } catch (SQLException e) {
            log.error("`{}`执行失败: {}", sql, e.getMessage(), e);
            throw e;
        }
    }

    public static List<String> schemas(final Connection connection) throws SQLException {
        final DatabaseMetaData it = connection.getMetaData();
        try (final ResultSet rt = it.getCatalogs()) {
            final List<String> schemas = new ArrayList<>();
            while (rt.next()) {
                schemas.add(rt.getString("TABLE_CAT"));
            }
            return schemas;
        } catch (SQLException e) {
            throw e;
        }
    }
}

File: mybatis-generator-gui/src/main/kotlin/com/wkit/lost/mybatis/generator/jdbc/JdbcTemplate.kt
package com.wkit.lost.mybatis.generator.jdbc

import com.alibaba.fastjson.JSON
import org.apache.logging.log4j.LogManager
import java.lang.Exception
import java.sql.Connection
import java.sql.PreparedStatement
import java.sql.ResultSet
import java.util.HashMap

/**
 * JDBC工具类
 * @author wvkity
 */
class JdbcTemplate {

    companion object {
        private val LOG = LogManager.getLogger(JdbcTemplate)

        /**
         * 查询操作
         * @param connection 数据库连接对象
         * @param sql SQL语句
         * @param args 参数
         */
        fun query(connection: Connection, sql: String, vararg args: Any?): MutableMap<String, Any?>? {
            val result = select(connection, sql, args)
            return if (result.isEmpty()) {
                null
            } else {
                result[0]
            }
        }

        /**
         * 查询操作
         * @param clazz 指定返回类型
         * @param connection 数据库连接对象
         * @param sql SQL语句
         * @param args 参数
         * @param <T> 泛型类型
         */
        fun <T> query(clazz: Class<T>, connection: Connection, sql: String, vararg args: Any?): T? {
            val result = query(connection, sql, *args)
            result?.run {
                if (this.isNotEmpty()) {
                    return JSON.parseObject(JSON.toJSONString(this), clazz)
                }
            }
            return null
        }

        /**
         * 查询操作
         * @param clazz 指定返回类型
         * @param connection 数据库连接对象
         * @param sql SQL语句
         * @param args 参数
         * @param <T> 泛型类型
         */
        fun <T> queryList(clazz: Class<T>, connection: Connection, sql: String, vararg args: Any): MutableList<T?>? {
            val result = queryList(connection, sql, *args)
            if (result.isNotEmpty()) {
                return JSON.parseArray(JSON.toJSONString(result), clazz)
            }
            return null
        }

        /**
         * 查询操作
         * @param connection 数据库连接对象
         * @param sql SQL语句
         * @param args 参数
         */
        fun queryList(connection: Connection, sql: String, vararg args: Any?): MutableList<MutableMap<String, Any?>> {
            return select(connection, sql, *args)
        }

        private fun select(connection: Connection, sql: String, vararg args: Any?): MutableList<MutableMap<String, Any?>> {
            var statement: PreparedStatement? = null
            var resultSet: ResultSet? = null
            val result = ArrayList<MutableMap<String, Any?>>()
            LOG.info("The SQL statement executed: $sql")
            try {
                statement = connection.prepareStatement(sql)
                if (args.isNotEmpty()) {
                    for ((i, v) in args.withIndex()) {
                        statement.setObject((i + 1), v)
                    }
                }
                resultSet = statement.executeQuery();
                while (resultSet.next()) {
                    val map = HashMap<String, Any?>()
                    val metadata = resultSet.metaData
                    for (i in 1..metadata.columnCount) {
                        val columnName = metadata.getColumnName(i)
                        map[columnName] = resultSet.getObject(columnName)
                    }
                    result.add(map)
                }
            } catch (e: Exception) {
                LOG.error("`{}`执行失败: {}", sql, e.message, e)
            } finally {
                close(resultSet)
                close(statement)
                close(connection)
            }
            return result
        }


        /**
         * 更新操作
         * @param connection 数据库连接对象
         * @param sql SQL语句
         * @param args 参数
         */
        fun update(connection: Connection, sql: String, vararg args: Any?): Int {
            var statement: PreparedStatement? = null
            LOG.info("The SQL statement executed: $sql")
            try {
                statement = connection.prepareStatement(sql)
                return if (args.isEmpty()) {
                    statement.executeUpdate()
                } else {
                    for ((i, v) in args.withIndex()) {
                        statement.setObject((i + 1), v)
                    }
                    statement.executeUpdate()
                }
            } catch (e: Exception) {
                LOG.error("`{}`执行失败: {}", sql, e.message, e)
                return 0
            } finally {
                close(statement)
                close(connection)
            }
        }

        fun close(resultSet: ResultSet?) {
            try {
                resultSet?.run { this.close() }
            } catch (e: Exception) {
                // ignore
            }
        }

        fun close(statement: PreparedStatement?) {
            try {
                statement?.run { this.close() }
            } catch (e: Exception) {
                // ignore
            }
        }

        fun close(connection: Connection?) {
            try {
                connection?.run { this.close() }
            } catch (e: Exception) {
                // ignore
            }
        }
    }
}